Rollback tests

adityabharadwaj198 opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
* feature


* **What is the current behavior?** (You can also link to an open issue here)
* Currently we have no rollback tests  


* **What is the new behavior (if this is a feature change)?**
* Functionality to do Rollback testing added 


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
Yes: https://github.com/marqo-ai/marqo/actions/runs/12047295783

* **Related Python client changes** (link commit/PR here)
NA

* **Related documentation changes** (link commit/PR here)
* NA


* **Other information**:


* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

adityabharadwaj198 committed:
Create backwards_compatibility_marqo_orchestrator.yml

adityabharadwaj198 committed:
Create backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Update backwards_compatibility_marqo_orchestrator.yml

changing to the right repo name

adityabharadwaj198 committed:
Update backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Update backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Update backwards_compatibility_marqo_execution.yml

printing some things

adityabharadwaj198 committed:
Update backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Changing secrets variables

adityabharadwaj198 committed:
Changing from AMD to ARM

adityabharadwaj198 committed:
changing from arm to amd

adityabharadwaj198 committed:
Updating backwards compatibility & rollback to use Start-Runner step

adityabharadwaj198 committed:
Adding workflow .yml files, scripts file for tox, and base test classes and tox.ini

adityabharadwaj198 committed:
Echoing workflow data from trigger execution workflows

adityabharadwaj198 committed:
Removing trailing whitespace from the curl command which invokes execution workflow

adityabharadwaj198 committed:
Removing whitespaces from =

adityabharadwaj198 committed:
Printing some variables

adityabharadwaj198 committed:
Update tox.ini - defined backwards compatibility tests in envlist

adityabharadwaj198 committed:
some debugging

adityabharadwaj198 committed:
making py3-docker_backwards_compatibility_tests to inherit from testenv

adityabharadwaj198 committed:
List tox environments

adityabharadwaj198 committed:
changing tox.ini file

adityabharadwaj198 committed:
Changing some more stuff

adityabharadwaj198 committed:
I can fix him

adityabharadwaj198 committed:
I can fix him

adityabharadwaj198 committed:
Committing compatibility test runner to view diff

adityabharadwaj198 committed:
Committing changes so I don't lose em

adityabharadwaj198 committed:
Committing all changes - will refine later

adityabharadwaj198 committed:
adding .yml file to build and push image to ecr

adityabharadwaj198 committed:
Merge branch 'marqo-ai:mainline' into mainline

adityabharadwaj198 committed:
Merge remote-tracking branch 'origin/mainline' into mainline

adityabharadwaj198 committed:
Changed the orchestrator yaml file to call adityabharadwaj198/marqo's build push image workflow

adityabharadwaj198 committed:
Merge branch 'marqo-ai:mainline' into mainline

adityabharadwaj198 committed:
Changing secrets variable name.

adityabharadwaj198 committed:
Adding logic to stop EC2 runner

adityabharadwaj198 committed:
Changing github actions tag

adityabharadwaj198 committed:
Changing how test execution happens from github actions

adityabharadwaj198 committed:
Changing image repo name for ECR

adityabharadwaj198 committed:
Added logic to push to ecr & pull from ecr. Also added logic to poll ecr before executing backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Removing unnecessary logs + changing from private repo to marqo-ai/marqo repo

adityabharadwaj198 committed:
Reverting changes to tox.ini file

adityabharadwaj198 committed:
Removing unncecessary comments and dead code

adityabharadwaj198 committed:
removing debug statements

adityabharadwaj198 committed:
Merge branch 'mainline' into mainline

adityabharadwaj198 committed:
Making workflow run on mainline push?

adityabharadwaj198 committed:
Merge remote-tracking branch 'adityabharadwaj198-origin/mainline' into mainline

adityabharadwaj198 committed:
Fixed the version generation logic

adityabharadwaj198 committed:
Changing repos back to point to adityabharadwaj198/marqo

adityabharadwaj198 committed:
Making orchestrate need poll-for-image-existence & fixing the generate_versions.py path

adityabharadwaj198 committed:
Making orchestrate need poll-for-image-existence & fixing the generate_versions.py path

adityabharadwaj198 committed:
Updating the value of sleep time & max retries

adityabharadwaj198 committed:
Updating the value of sleep time & max retries

adityabharadwaj198 committed:
Fixing perf_tests/README.md

adityabharadwaj198 committed:
Fixing perf_tests/README.md

adityabharadwaj198 committed:
Fixing backwards_compatibility_marqo_execution.yml by changing needs

adityabharadwaj198 committed:
Fixing the path to compatibility_test_runner.py from backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Fixing the path to compatibility_test_runner.py from backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Adding a step to ls

adityabharadwaj198 committed:
Adding a step to ls -la

adityabharadwaj198 committed:
Adding a step to ls -lR

adityabharadwaj198 committed:
checking out adityabharadwaj198/marqo branch

adityabharadwaj198 committed:
Adding logic to install requirements

adityabharadwaj198 committed:
Adding logic to install pip-tools which installs pip-compile

adityabharadwaj198 committed:
Adding semver to requirements.in

adityabharadwaj198 committed:
Adding logic to make from_image & to_image as optional arguments

adityabharadwaj198 committed:
Merge branch 'marqo-ai:mainline' into mainline

adityabharadwaj198 committed:
Changing how we execute test_runner.py

adityabharadwaj198 committed:
Making changes for easier debugging

adityabharadwaj198 committed:
Changing runs on for backwards_compatibility job

adityabharadwaj198 committed:
Removing logic to resize disk

adityabharadwaj198 committed:
Uncommenting things

adityabharadwaj198 committed:
Removing step to list contents of directory and adding step to login to ECR registry

adityabharadwaj198 committed:
better exception handling when pulling docker from ECR

adityabharadwaj198 committed:
Adding some debug statements and running pytest with the directory specified

adityabharadwaj198 committed:
Merge branch 'refs/heads/aditya/compatibility-tests' into mainline

# Conflicts:
#	.github/workflows/backwards_compatibility_marqo_execution.yml

adityabharadwaj198 committed:
Merge branch 'marqo-ai:mainline' into mainline

adityabharadwaj198 committed:
Added logic to run general compatibility tests that were ran pre 2.13 release.

adityabharadwaj198 committed:
Removed some debugging steps, bumped up python version, and changed the to_version determination logic.

adityabharadwaj198 committed:
Moved generate_versions.py to /scripts/generate_version.py, and removed 2 TODOs related to using / determining to_version

adityabharadwaj198 committed:
Adding logic to check image existence (in case of a prerun) and only then execute build_push_img.yml & removing some debug steps

adityabharadwaj198 committed:
Removing unused code etc

adityabharadwaj198 committed:
Removing usage of pip-compile

adityabharadwaj198 committed:
Added tons of documentation

adityabharadwaj198 committed:
Revision # 1 addressing reviewer comments

adityabharadwaj198 committed:
Merge branch 'mainline' into mainline

adityabharadwaj198 committed:
Edited an unclear comment

adityabharadwaj198 committed:
Revision # 2 - Addressing comments

adityabharadwaj198 committed:
Making execution workflow callable

adityabharadwaj198 committed:
Changing run-execution-workflow

adityabharadwaj198 committed:
Changing some more things

adityabharadwaj198 committed:
Fixing

adityabharadwaj198 committed:
Adding inputs to workflow_call

adityabharadwaj198 committed:
Started invoking local workflows, and changed logic to start using image digest instead of commit hash

adityabharadwaj198 committed:
Added logic to do full test run, changed print to debug log statements and fixed a bunch of testing related things

adityabharadwaj198 committed:
Made workflow use image digest, and addressed various comments

adityabharadwaj198 committed:
Added a bunch of todos and addressed some more comments

adityabharadwaj198 committed:
Added 1 more todo

adityabharadwaj198 committed:
Incorporating comments regarding being able to use image tag & digest both

adityabharadwaj198 committed:
Fixing small errors in compatibility_test_runner.py

adityabharadwaj198 committed:
Making TO_VERSION_IMAGE_IDENTIFIER into a variable $TO_VERSION_IMAGE_IDENTIFIER

adityabharadwaj198 committed:
Set image_identifier as output of build_push_img.yml

adityabharadwaj198 committed:
changed a print statement to logger.debug, removed some debugging logic and removed some logs which were irrelevant

adityabharadwaj198 committed:
Making it work even when image already exists

adityabharadwaj198 committed:
Fixing test cases

adityabharadwaj198 committed:
Removed comments

adityabharadwaj198 committed:
Merge branch 'mainline' into mainline

adityabharadwaj198 committed:
Ignoring tests/backwards_compatibility_tests folder from large model unit tests, 200gb unit tests

adityabharadwaj198 committed:
Merge remote-tracking branch 'adityabharadwaj198-origin/mainline' into mainline

adityabharadwaj198 committed:
Changing how we import compatibility_test_logger

adityabharadwaj198 committed:
Merge remote-tracking branch 'marqo-ai/marqo-origin-don't-push-to-this-origin-from-this-folder/mainline' into mainline

# Conflicts:
#	.github/workflows/backwards_compatibility_marqo_execution.yml
#	.github/workflows/backwards_compatibility_marqo_orchestrator.yml
#	.github/workflows/largemodel_unit_test_CI.yml
#	tests/backwards_compatibility_tests/compatibility_test_runner.py
#	tests/backwards_compatibility_tests/conftest.py

adityabharadwaj198 committed:
All changes required for running rollback tests

adityabharadwaj198 committed:
Renamed .yml files to make them sound generic for bw-compatibility & rollback both

adityabharadwaj198 committed:
Fixing indentation when running tests and renaming job name to make it sound generic

adityabharadwaj198 committed:
Revert "Renamed .yml files to make them sound generic for bw-compatibility & rollback both"

This reverts commit 0c78c4621c90a3f1e6c881dc7d6ed530fd67981e.

adityabharadwaj198 committed:
Changing if condition and removing one log statement

papa99do commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -625,10 +717,23 @@ def copy_state_from_container(
+        logger.error(f"from Major version {from_major_version} & to Major version {to_major_version} cannot "
+                     f"be used for running backwards compatibility tests or rollback tests"
+                     f"since they are from different major versions")
+        sys.exit(0)
```
when validation errors out, should we exit with an error code to indicate the test fails to run? Or maybe just throw an exception?

adityabharadwaj198 replied:
oops, meant to use a sys.exit(1) here - thanks for catching this. I'll change it to sys.exit(1). I also explored the possibility of throwing an exception here, it will work but it may produce noisy logs which may include python stack trace etc. So I've made changes such that this will throw an exception and the exception will be caught, logged and sys.exit(1) will be called

farshidz commented on .github/workflows/backwards_compatibility_marqo_execution.yml:
```diff
@@ -116,18 +116,34 @@ jobs:
+            --to_version_image_identifier "$TO_VERSION_IMAGE_IDENTIFIER" \
+
+      # Step to run rollback tests. Only runs if backwards compatibility tests have succeeded
+      - name: Run rollback tests
```
Running these sequentially seems unnecessary and inefficient

adityabharadwaj198 replied:
Rollback tests' initial tests are the same  as backwards compatibility tests (https://www.notion.so/Backwards-compatibility-and-rollback-testing-16d64207c72e486fa25001192f20817b?pvs=4#0acae9b3b5024b4b946e373a30f02c68). The idea here was that if anything is failing during the steps which are in both backwards_compatibility & rollback & they failing in backwards_compatibility why run rollback at all.

farshidz replied:
But if they're going to pass, they'll pass faster. Or if they're going to fail for different reasons. Slow tests are a major issue for us currently

adityabharadwaj198 replied:
made them run in parallel

farshidz commented on .github/workflows/backwards_compatibility_marqo_execution.yml:
```diff
@@ -116,18 +116,34 @@ jobs:
+        env:
+          FROM_VERSION: ${{ inputs.from_version || github.event.inputs.from_version }}
+          TO_VERSION: ${{ inputs.to_version || github.event.inputs.to_version }}
+          TO_VERSION_IMAGE_IDENTIFIER: ${{ inputs.to_version_image_identifier || github.event.inputs.to_version_image_identifier }}
```
I really think to_image will make more sense but perhaps seek a second opinion. Name looks unnecessarily long

adityabharadwaj198 replied:
I have changed it to to_image wherever it made sense to do so, I've also shortened the variable name where-ever I couldn't change it to to_image, hope that helps

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -212,35 +207,46 @@ def start_marqo_from_version_container(from_version: str, from_version_volume, e
-    to_version = semver.VersionInfo.parse(to_version)
-    from_version = semver.VersionInfo.parse(from_version)
+        f"Starting Marqo container with target version: {target_version}, "
+        f"source version: {source_version} "
```
Are we being consistent with the terminology? We shouldn't use both source version and from version especially in input args and outputs

adityabharadwaj198 replied:
Using target_version / source_version in this method makes more sense because this method is common to both backwards compatibility & rollback. In backwards compatibility it's required to upgrade marqo version (so target_version will be to_version and source_version will be from_version), and in rollback tests it's used to upgrade marqo version + downgrade marqo version (in which case target_version will be from_version and source_version will be to_version since you're downgrading). Here upgrading and downgrading basically mean running a newer marqo version & running an older marqo version resp.

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -254,24 +260,22 @@ def start_marqo_to_version_container(to_version: str, from_version: str, from_ve
-
 
-    if from_version >= marqo_transfer_state_version and to_version >= marqo_transfer_state_version:
+    if source_version >= marqo_transfer_state_version and target_version >= marqo_transfer_state_version:
```
Likewise with target version. Are we consistent with terminology?

adityabharadwaj198 replied:
^

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -615,6 +525,188 @@ def copy_state_from_container(
+
+        #Start from_version container
+        start_marqo_container(from_version, from_version_volume)
+        logger.debug(f"Started marqo container {from_version}")
```
Marqo should be capitalized

adityabharadwaj198 replied:
fixed everywhere

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -615,6 +525,188 @@ def copy_state_from_container(
+        try:
+            run_tests_across_versions("prepare", from_version)
+        except Exception as e:
+            logger.error(f"Error running tests across versions in 'prepare' mode: {e}")
```
Logging and raising in an antipattern

farshidz replied:
Applies to all other cases in this PR

adityabharadwaj198 replied:
Fixed everywhere

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -615,6 +525,188 @@ def copy_state_from_container(
+        logger.debug(f"Started marqo to_version: {to_version} container by transferring state")
+        # Step 4: Run tests
+        try:
+            run_tests_across_versions("test", from_version)
```
Can we have an enum for mode? Passing strings around is error prone and confusing

adityabharadwaj198 replied:
done

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -366,130 +370,35 @@ def cleanup_volumes():
+    run_test_mode(from_version)
+    full_test_run(to_version)
+
+def full_test_run(marqo_version: str):
```
When we run prepare again here, do we use new index names?

adityabharadwaj198 replied:
No we don't need to do that because after running prepare 1st time (and consequently having run test also), we will delete the indexes. We will then run prepare + test here again

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -366,130 +370,35 @@ def cleanup_volumes():
-    Raises:
-        ValueError: If the major versions of from_version and to_version are incompatible.
-        Exception: If there is an error during the test process.
+def run_tests_across_versions(mode: str, from_version: str):
```
Both name and description appear misleading because this method doesn't actually perform state transfer and is operating on a single version of marqo only

adityabharadwaj198 replied:
Changed name & description

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -615,6 +525,188 @@ def copy_state_from_container(
+
+        # Step 1: Start a Marqo container using from_version and run tests in prepare mode.
+        start_marqo_container(from_version, from_version_volume)
+        logger.debug(f"Started marqo container {from_version}")
```
nit I'm on the fence with these log levels. They feel like info messages. I also see you had to change your default log level to debug

adityabharadwaj198 replied:
I've changed some logs to info, kept some logs as debug level (which we don't need to see in test runs) . And have changed the default log level to info. So anyone whose developing can change them to debug level and see all the logs (debug + info).

adityabharadwaj198 committed:
Addressing comments on rollback PR

farshidz commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -611,25 +527,228 @@ def copy_state_from_container(
+        logger.info(f"Started Marqo container {from_version}")
+
+        try:
+            run_tests_in_mode(Mode.PREPARE.value, from_version)
```
Why do you need `.value`? Your func can just expect the enum

adityabharadwaj198 replied:
changed it so it accepts enum

adityabharadwaj198 committed:
Addressing comments on rollback PR # 2

1. Made rollback & bw compatibility run in parallel
2. Not logging full exception anymore if I'm raising it
3. removed pillow and numpy from requirements

adityabharadwaj198 committed:
Changing job names

adityabharadwaj198 committed:
Changing mode's type to string & remvoing options parameter from it

adityabharadwaj198 committed:
Changing exception handling

papa99do commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -221,36 +226,36 @@ def start_marqo_container_by_transferring_state(target_version: str, source_vers
         subprocess.run(["docker", "rm", "-f", container_name], check=True)
     except subprocess.CalledProcessError:
-        logger.debug(f"Container {container_name} not found, skipping removal.")
+        logger.warn(f"Container {container_name} not found, skipping removal.")
```
nit. use warning instead, warn is deprecated.

papa99do commented on tests/backwards_compatibility_tests/requirements.txt:
```diff
@@ -1,6 +1,4 @@
 pytest==7.4.3
-pillow==9.3.0
-numpy==1.23.4
 marqo
```
we do not specify marqo version here, which means we always use the latest one. Would this cause issue since py-marqo has requirements of minimum version of Marqo server. Can we use newer version of py-marqo to send request to older version of Marqo?

adityabharadwaj198 replied:
discussed offline that I have tested till version 2.9 with latest py-marqo

adityabharadwaj198 committed:
Changing logger.warn to logger.warning

adityabharadwaj198 committed:
Merge branch 'mainline' into aditya/rollback
